pkg/controller/podgroup.go

```diff
@@ -0,0 +1,168 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"sort"
+
+	"github.com/google/go-cmp/cmp"
+	common "github.com/kubeflow/common/pkg/apis/common/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog"
+	volcanov1beta1 "volcano.sh/apis/pkg/apis/scheduling/v1beta1"
+
+	kubeflow "github.com/kubeflow/mpi-operator/pkg/apis/kubeflow/v2beta1"
+)
+
+type replicaPriority struct {
+	priority int32
+
+	common.ReplicaSpec
+}
+
+// newPodGroup creates a new PodGroup for an MPIJob
+// resource. It also sets the appropriate OwnerReferences on the resource so
+// handleObject can discover the MPIJob resource that 'owns' it.
+func (c *MPIJobController) newPodGroup(mpiJob *kubeflow.MPIJob) *volcanov1beta1.PodGroup {
+	minMember := workerReplicas(mpiJob) + 1
+	queueName := mpiJob.Annotations[volcanov1beta1.QueueNameAnnotationKey]
+	var minResources *corev1.ResourceList
+	if schedulingPolicy := mpiJob.Spec.RunPolicy.SchedulingPolicy; schedulingPolicy != nil {
+		if schedulingPolicy.MinAvailable != nil {
+			minMember = *schedulingPolicy.MinAvailable
+		}
+		if len(schedulingPolicy.Queue) != 0 {
+			queueName = schedulingPolicy.Queue
+		}
+		if schedulingPolicy.MinResources != nil {
+			minResources = schedulingPolicy.MinResources
+		}
+	}
+	if minResources == nil {
+		minResources = c.calcPGMinResources(minMember, mpiJob.Spec.MPIReplicaSpecs)
+	}
+	return &volcanov1beta1.PodGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      mpiJob.Name,
+			Namespace: mpiJob.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(mpiJob, kubeflow.SchemeGroupVersionKind),
+			},
+		},
+		Spec: volcanov1beta1.PodGroupSpec{
+			MinMember:         minMember,
+			Queue:             queueName,
+			PriorityClassName: calcPriorityClassName(mpiJob.Spec.MPIReplicaSpecs, mpiJob.Spec.RunPolicy.SchedulingPolicy),
+			MinResources:      minResources,
+		},
+	}
+}
+
+func calcPriorityClassName(
+	replicas map[kubeflow.MPIReplicaType]*common.ReplicaSpec,
+	schedulingPolicy *kubeflow.SchedulingPolicy,
+) string {
+	if schedulingPolicy != nil && len(schedulingPolicy.PriorityClass) != 0 {
+		return schedulingPolicy.PriorityClass
+	}
+	if l := replicas[kubeflow.MPIReplicaTypeLauncher]; l != nil && len(l.Template.Spec.PriorityClassName) != 0 {
+		return l.Template.Spec.PriorityClassName
+	} else if w := replicas[kubeflow.MPIReplicaTypeWorker]; w != nil && len(w.Template.Spec.PriorityClassName) != 0 {
+		return w.Template.Spec.PriorityClassName
+	}
+	return ""
+}
+
+func (c *MPIJobController) calcPGMinResources(
```

Comment:
add a comment for what's happening here

Reply:
Still waiting.

What does volcano use that field? If it's all added up, how does it know whether they are consumed by one or multiple pods?

Reply:
> What does volcano use that field? If it's all added up, how does it know whether they are consumed by one or multiple pods?

I tried to dive into the volcano. The volcano has the Queue resource like ClusterQueue in Kueue. 
If Queue doesn't have enough resources requested by `minResources`  in PodGroup, volcano-scheduler doesn't schedule all Pods to Nodes.

Does that make sense?

Reply:
makes sense.... it looks like a quota system just like kueue. I wonder if they would be interested in reviewing this PR. Do you know anybody?

Reply:
I remember @zw0610 is familiar with volcano. And @zw0610 is a reviewer in this repository :)

Reply:
IMO, as a result of investigating the volcano, maybe we don't need the `calcPGMinResources` function since `minResources` don't affect gang-scheduling. And we may be able to propagate `schedulingPolicy.minResources` to PodGroup.

Reply:
@zw0610 Let us know what you think.

Reply:
Maybe, @shinytang6 is also interested in this PR.
cc: @shinytang6

Reply:
@alculquicondor They don't seem to have enough bandwidth to review this PR.
So, I'm thinking of removing `func calcPGMinResources` and then passing `SchedulingPolicy.MinResources` to `PodGroup.Spec.MinResources`.

Also, it would be good to keep tracking this.

Reply:
SGTM, we don't want to introduce functionality we aren't sure about.

Reply:
> we don't want to introduce functionality we aren't sure about.

I agree.

Reply:
Done.